src/main/java/org/cru/cas/client/integration/CopyCasAttributesToHeadersFilter.java

```diff
@@ -0,0 +1,251 @@
+        String mapping = filterConfig.getInitParameter(ATTRIBUTE_MAPPING_PARAMETER);
+        if (mapping != null && !mapping.isEmpty()) {
+            this.attributeMapping = parseAttributeMapping(mapping);
+        } else {
+            this.attributeMapping = new AttributeMapping() {
+                @Override
+                public boolean headerIsMapped(String headerName) {
+                    // case-insensitive "startsWith"
+                    return headerName.regionMatches(true, 0, CAS_PREFIX, 0, CAS_PREFIX.length());
+                }
+
+                @Override
+                public String getAttributeNameForHeader(String headerName) {
+                    return headerName.substring(CAS_PREFIX.length());
+                }
+
+                @Override
+                public Set<String> getHeaderNamesForAttribute(String attributeName) {
+                    return Collections.singleton(CAS_PREFIX + attributeName);
+                }
+            };
+        }
+    }
+
+    private AttributeMapping parseAttributeMapping(String mapping) {
+        Map<String, String> headersToAttributeNames = parseHeadersToAttributeNames(mapping);
+        Map<String, Set<String>> attributeNamesToHeaders = reverseMap(headersToAttributeNames);
+
+        return new AttributeMapping() {
+            @Override
+            public boolean headerIsMapped(String headerName) {
+                return headersToAttributeNames.containsKey(headerName);
+            }
+
+            @Override
+            public String getAttributeNameForHeader(String headerName) {
+                return headersToAttributeNames.get(headerName);
+            }
+
+            @Override
+            public Set<String> getHeaderNamesForAttribute(String attributeName) {
+                Set<String> headerNames = attributeNamesToHeaders.get(attributeName);
+                if (headerNames != null) {
+                    return headerNames;
+                } else {
+                    return Collections.emptySet();
+                }
+            }
+        };
+    }
+
+    private Map<String, String> parseHeadersToAttributeNames(String mapping) {
+        return Stream.of(mapping.split("\\s"))
+                .filter(string -> !string.isEmpty())
+                .peek(string -> {
+                    if (!string.contains(LINE_SEPARATOR)) {
+                        throw new IllegalArgumentException("bad mapping line: " + string);
+                    }
+                })
+                .collect(Collectors.toMap(
+                    line -> line.split(LINE_SEPARATOR)[1],
+                    line -> line.split(LINE_SEPARATOR)[0],
+                    throwingMergeFunction(),
+                    () -> new TreeMap<>(String.CASE_INSENSITIVE_ORDER)
+                ));
+    }
+
+    // borrowed from Collections.throwingMerger(), which is private
+    private <T> BinaryOperator<T> throwingMergeFunction() {
```

Comment:
Did you mean to name this `throwingMergerFunction()`?

Reply:
No, not really. The `Collectors` source code uses the parameter name `mergeFunction`, IIRC. I used that term.

Reply:
btw, if I had to do this again, I might follow this guy's [advice](https://stackoverflow.com/a/45210944/1174749) instead of using the four-arg Collectors.toMap().
But I'm not switching to that, now.